File: lib/models/project.dart
import 'package:model_interfaces/model_interfaces.dart';

import 'external_url.dart';

class Project implements WithIdTitle<String> {
  @override
  final String id;

  @override
  final String title;

  final String description;
  final int? year;
  final List<ExternalUrl> urls;
  final String imageUrl;
  final List<String> tags;
  final DateTime? dateTime;
  final String body;

  Project({
    required this.id,
    required this.title,
    required this.description,
    required this.year,
    required this.urls,
    required this.imageUrl,
    required this.tags,
    required this.dateTime,
    required this.body,
  });

  // TODO: Safe parsing. Return null and print error.
  factory Project.fromIdAndMap(String id, Map<String, dynamic> map) {
    return Project(
      id: id,
      title: map['title'],
      description: map['description'],
      year: map['year'],
      urls: ExternalUrl.fromMaps(map['urls'] ?? []),
      imageUrl: map['imageUrl'],
      tags: map['tags'].cast<String>(),
      dateTime: map['dateTime']?.toDate(),
      body: map['body'] ?? '',
    );
  }
}

File: firebase/functions/src/interfaces/Project.ts
import {StringObject, StringBoolObject} from "./maps";
import {Url} from "./Url";
import {firestore} from "firebase-admin";
import Timestamp = firestore.Timestamp;

export interface Project {
    id: string;
    body: string | undefined;
    dateTime: Date | undefined;
    description: string;
    imageUrl: string;
    keywords: string[];
    tags: string[];
    tagsMap: Map<string, boolean>;
    title: string;
    urls: Url[];
    year: number | undefined;
}

export function fromIdAndMap(id: string, map: StringObject): Project {
    const result = map as Project;

    result.id = id;
    result.dateTime = toDate(map['dateTime']);
    result.keywords = result.keywords ?? [];
    result.tags = result.tags ?? [];
    result.tagsMap = new Map<string, boolean>(Object.entries(map['tagsMap'] ?? {}));
    result.urls = result.urls ?? [];

    return result;
}

function toDate(value: any): Date | undefined {
    if (value instanceof Date) return value;
    if (value instanceof Timestamp) return value.toDate();
    if (typeof value === 'string') return new Date(value);

    return undefined;
}

export function getProjectKeywords(obj: Project): string[] {
    const result: string[] = [];

    result.push(...getTagKeywords(obj.tags));

    return result;
}

function getTagKeywords(tags: string[]): string[] {
    return tags.map(tag => `tag_${tag}`);
}

export function getProjectTagsMap(obj: Project): StringBoolObject {
    const result = {} as StringBoolObject;

    for (const tag of obj.tags) {
        result[tag] = true;
    }

    return result;
}
